guillotina/fields/annotation.py

```diff
@@ -0,0 +1,203 @@
+from guillotina import configure
+from guillotina import schema
+from guillotina.annotations import AnnotationData
+from guillotina.component import query_adapter
+from guillotina.interfaces import IAnnotations
+from guillotina.interfaces import IContentBehavior
+from guillotina.json.exceptions import ValueDeserializationError
+from guillotina.schema.interfaces import IField
+from guillotina.schema.interfaces import IPatchFieldOperation
+from zope.interface import implementer
+
+
+_default = object()
+
+
+class BucketListValue:
+
+    def __init__(self, annotation_prefix='bucketlist-', bucket_len=5000):
+        self.current_annotation_index = 0
+        self.annotations_metadata = {}
+        self.annotation_prefix = annotation_prefix
+        self.bucket_len = bucket_len
+
+    def get_annotation_name(self, index):
+        return f'{self.annotation_prefix}{index}'
+
+    async def get_annotation(self, context, annotation_index, create=True):
+        annotation_name = self.get_annotation_name(annotation_index)
+        annotations_container = IAnnotations(context)
+        annotation = annotations_container.get(annotation_name, _default)
+        if annotation is _default:
+            annotation = await annotations_container.async_get(
+                annotation_name, _default)
+        if annotation is _default:
+            if not create:
+                return
+            # create
+            annotation = AnnotationData()
+            annotation.update({
+                'items': []
+            })
+            await annotations_container.async_set(annotation_name, annotation)
+        if annotation_index not in self.annotations_metadata:
+            self.annotations_metadata[annotation_index] = {}
+        return annotation
+
+    async def append(self, context, value):
+        if self.current_annotation_index in self.annotations_metadata:
+            metadata = self.annotations_metadata[self.current_annotation_index]
+            if metadata.get('len', 0) >= self.bucket_len:
+                # split here
+                self.current_annotation_index += 1
+
+        annotation = await self.get_annotation(context, self.current_annotation_index)
+        metadata = self.annotations_metadata[self.current_annotation_index]
+        metadata['len'] = metadata.get('len', 0) + 1
+        annotation['items'].append(value)
+        annotation._p_register()
+
+    async def extend(self, context, value):
+        for item in value:
+            await self.append(context, item)
+
+    def __len__(self):
+        total = 0
+        for metadata in self.annotations_metadata.values():
```

Comment:
If the amount of buckets is really large we may have a lot of loads to get the total?

Reply:
This is a small metadata object. If you have 1 million buckets, it might be slow but if you're using this to store billions of items(bucket size * # of buckets), you're doing something wrong...

Reply:
Can you approve?